lib/src/main/java/graphql/nadel/Nadel.kt

```diff
@@ -75,75 +99,79 @@ class Nadel private constructor(
         return try {
             val executionInstrumentation = instrumentation.beginQueryExecution(instrumentationParameters)
 
-            parseValidateAndExecute(executionInput, querySchema, instrumentationState, nadelExecutionParams)
-                // finish up instrumentation
-                .whenComplete { result: ExecutionResult?, t: Throwable? ->
-                    executionInstrumentation.onCompleted(result, t)
-                }
-                .exceptionally { throwable: Throwable ->
-                    if (throwable is AbortExecutionException) {
-                        throwable.toExecutionResult()
-                    } else if (throwable is CompletionException && throwable.cause is AbortExecutionException) {
-                        val abortException = throwable.cause as AbortExecutionException
-                        abortException.toExecutionResult()
-                    } else if (throwable is RuntimeException) {
-                        throw throwable
-                    } else {
-                        throw RuntimeException(throwable)
+            val result = try {
+                parseValidateAndExecute(executionInput, instrumentationState, hints)
+                    .also {
+                        executionInstrumentation.onCompleted(it, null)
```

Comment:
`onCompleted` was being called from `whenComplete` before. Is calling it from `also` going to result in it executing right after the function returns but before the future resolves?

Reply:
Also: do we have test coverage that asserts these instrumentation methods are being call as expected?

Reply:
So if I understand `whenComplete` then this is the exact same behavior.

`.also` isn't some special function. This code is exactly the same as
 
```
val temp1 = parseValidateAndExecute(executionInput, instrumentationState, hints)
executionInstrumentation.onCompleted(temp1, null)
temp1 // return
```

Here's the code definition for `.also`

```
public inline fun <T> T.also(block: (T) -> Unit): T {
    block(this)
    return this
}
```

Reply:
I'll check on the tests.

Reply:
Ah, I see. So calling a suspended function will actually block the execution. 
Whereas this
```
async {
parseValidateAndExecute(executionInput, instrumentationState, hints)
}
```
would return a `Defer` which is kind of like a CF in Java

Reply:
Ok I introduced some more tests here https://github.com/atlassian-labs/nadel/pull/539